### PR TITLE
 Applications: Add new KeyboardSettings application

### PR DESCRIPTION
**Pull request by asliturk**
Add new app for keyboard setting to support GUI and CLI
CLI:
```
# KeyboardSettings trq
```
GUI:
Screenshot from 2020-06-04 15-11-45

### CODE REVIEW

File: Kernel/Process.cpp [hunk 1]
```diff
@@ -4290,7 +4290,7 @@ int Process::sys$setkeymap(const Syscall::SC_setkeymap_params* user_params)
     if (!is_superuser())
         return -EPERM;
 
-    REQUIRE_NO_PROMISES;
+    REQUIRE_PROMISE(stdio);
```
Comment (bugaevc) [inline Kernel/Process.cpp:4293]: Disagree. Changing the system keymap is not something you'd want a process that has only promised stdio to be able to do.

Generally, stdio means only doing I/O on the file descriptors you already have opened (capability-style), manipulating your own memory mappings, and some other basic stuff like getting your PID. Here's a simple test you could use to decide if some functionality fits into stdio: can it be used to impact the rest of the system in any way, not counting any pre-opened files? If so, it's a bad fit for stdio.

File: Applications/KeyboardSettings/CharacterMapFileListModel.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+#pragma once
+
+#include <AK/NonnullRefPtr.h>
+#include <AK/Vector.h>
+#include <LibGUI/Model.h>
+
+template<typename T>
+class CharacterMapFileListModel final : public GUI::Model {
```
Comment (bugaevc) [inline Applications/KeyboardSettings/CharacterMapFileListModel.h:34]: Any reason for this to be a template?

File: Applications/KeyboardSettings/CharacterMapFileListModel.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+//    {
+//        return { 70, Gfx::TextAlignment::CenterLeft };
+//    }
+
+    virtual GUI::Variant data(const GUI::ModelIndex& index, Role role = Role::Display) const override
+    {
+        if (role == Role::Display)
+            return m_data.at(index.row());
+
+        return {};
```
Comment (bugaevc) [inline Applications/KeyboardSettings/CharacterMapFileListModel.h:65]: If you make this a non-template, these all bodies should move out-of-line.

File: Applications/KeyboardSettings/CharacterMapFileListModel.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+//    virtual ColumnMetadata column_metadata(int) const override
+//    {
+//        return { 70, Gfx::TextAlignment::CenterLeft };
+//    }
+
+    virtual GUI::Variant data(const GUI::ModelIndex& index, Role role = Role::Display) const override
+    {
+        if (role == Role::Display)
```
Comment (bugaevc) [inline Applications/KeyboardSettings/CharacterMapFileListModel.h:62]:
```cpp
ASSERT(index.is_valid());
ASSERT(index.column() == 0);
```

File: Applications/KeyboardSettings/CharacterMapFileListModel.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        did_update();
+    }
+
+private:
+    explicit CharacterMapFileListModel(Vector<T>& data)
+        : m_data(data)
+    {
+    }
+
+    Vector<T>& m_data;
```
Comment (bugaevc) [inline Applications/KeyboardSettings/CharacterMapFileListModel.h:79]: What kind of "data" is this? To me, it sounds like a list of paths to files containing keyboard maps. Perhaps name it something like that then :slightly_smiling_face:

File: Applications/KeyboardSettings/main.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+int main(int argc, char** argv)
+{
+    if (pledge("accept cpath rpath exec fattr proc shared_buffer stdio thread unix", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    const char* path = nullptr;
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(path, "The mapping file to be used.", "file", Core::ArgsParser::Required::No);
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:54]: No period at the end of the usage string, please.

File: Applications/KeyboardSettings/main.cpp [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    args_parser.parse(argc, argv);
+
+    if (path != nullptr) {
+        Keyboard::CharacterMap character_map(path);
+        int ret_val = character_map.set_system_map();
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:59]:
```suggestion
        int rc = character_map.set_system_map();
```

File: Applications/KeyboardSettings/main.cpp [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+        if (ret_val == -EPERM)
+            fprintf(stderr, "Permission denied.\n");
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:62]: And what if it's another error?

File: Applications/KeyboardSettings/main.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+        if (ret_val == 0)
+            fprintf(stderr, "New keymap loaded from \"%s\".\n", path);
+
+        return 0;
+    }
+
+    // If there is no cmd line parameter go for GUI.
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:70]: Either "cmdline" or "command line" (or "command-line").

File: Applications/KeyboardSettings/main.cpp [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+    auto app_icon = GUI::Icon::default_icon("app-keyboard-settings");
+
+    Vector<String> character_map_files;
+    Core::DirIterator iterator("/res/keymaps/", Core::DirIterator::Flags::SkipDots);
+    while (iterator.has_next()) {
+        auto name = iterator.next_path();
+        name.replace(".json", "");
+        character_map_files.append(name);
+    }
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:86]: You forgot to handle the error case (`iterator.has_error()`).

File: Applications/KeyboardSettings/main.cpp [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    root_widget.set_fill_with_background_color(true);
+    root_widget.layout()->set_spacing(0);
+    root_widget.layout()->set_margins({ 4, 4, 4, 4 });
+
+    auto& character_map_file_selection_container = root_widget.add<GUI::Widget>();
+    character_map_file_selection_container.set_layout<GUI::HorizontalBoxLayout>();
+    character_map_file_selection_container.set_size_policy(GUI::SizePolicy::Fill, GUI::SizePolicy::Fixed);
+    character_map_file_selection_container.set_preferred_size(0, 22);
+
+    auto& character_map_filelabel = character_map_file_selection_container.add<GUI::Label>();
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:105]:
```suggestion
    auto& character_map_file_label = character_map_file_selection_container.add<GUI::Label>();
```

File: Applications/KeyboardSettings/main.cpp [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+    character_map_file_combo.set_preferred_size(0, 22);
+    character_map_file_combo.set_only_allow_values_from_model(true);
+    character_map_file_combo.set_model(*CharacterMapFileListModel<AK::String>::create(character_map_files));
+
+    root_widget.layout()->add_spacer();
+
+    auto apply_settings = [&](bool quit) {
+        String character_map_file = character_map_file_combo.text();
+        if (character_map_file.is_empty()) {
+            GUI::MessageBox::show("Please select cahracter mapping file.", "Keyboard settings", GUI::MessageBox::Type::Error, GUI::MessageBox::InputType::OK, window);
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:122]:
```suggestion
            GUI::MessageBox::show("Please select character mapping file.", "Keyboard settings", GUI::MessageBox::Type::Error, GUI::MessageBox::InputType::OK, window);
```

File: Applications/KeyboardSettings/CharacterMapFileListModel.h (hunk 5 above)
Comment (asliturk) [reply Applications/KeyboardSettings/CharacterMapFileListModel.h:79]: You are right, it is file_name vector. How is `m_file_names`?

Comment (bugaevc) [reply Applications/KeyboardSettings/CharacterMapFileListModel.h:79]: That, or `m_map_names`

File: Kernel/Process.cpp (hunk 1 above)
Comment (asliturk) [reply Kernel/Process.cpp:4293]: Hi Sergey,
When program sets pledge("x") REQUIRE_NO_PROMISES raise error. For me the best is stdio :)

Comment (bugaevc) [reply Kernel/Process.cpp:4293]: That is the point of `REQUIRE_NO_PROMISES`, yes :) we really don't want sandboxed programs messing with system keymap, do we?

Comment (asliturk) [reply Kernel/Process.cpp:4293]: I roll back it to `REQUIRE_NO_PROMISESI`

File: Applications/KeyboardSettings/CharacterMapFileListModel.h [hunk 13]
```diff
@@ -24,26 +24,51 @@
...
-{
-    if (argc != 2) {
-        fprintf(stderr, "usage: keymap <name|file>\n");
-        return 0;
+#include <AK/Vector.h>
+#include <LibGUI/Model.h>
+
+class CharacterMapFileListModel final : public GUI::Model {
+public:
+    static NonnullRefPtr<CharacterMapFileListModel> create(Vector<String>& data)
```
Comment (bugaevc) [inline Applications/KeyboardSettings/CharacterMapFileListModel.h:34]:
```suggestion
    static NonnullRefPtr<CharacterMapFileListModel> create(Vector<String>& file_names)
```

File: Applications/KeyboardSettings/CharacterMapFileListModel.h [hunk 14]
```diff
@@ -24,26 +24,51 @@
...
 
-    if (ret_val == 0)
-        fprintf(stderr, "New keymap loaded from \"%s\".\n", filename.characters());
+    virtual void update() override
+    {
+        did_update();
+    }
+
+private:
+    explicit CharacterMapFileListModel(Vector<String>& data)
```
Comment (bugaevc) [inline Applications/KeyboardSettings/CharacterMapFileListModel.h:68]:
```suggestion
    explicit CharacterMapFileListModel(Vector<String>& file_names)
```

File: Applications/KeyboardSettings/main.cpp [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    args_parser.parse(argc, argv);
+
+    if (path != nullptr) {
+        Keyboard::CharacterMap character_map(path);
+        int rc = character_map.set_system_map();
+
+        if (rc == -EPERM)
+            fprintf(stderr, "Permission denied.\n");
+        else if (rc != 0)
+            fprintf(stderr, "Unknown error.\n");
```
Comment (bugaevc) [inline Applications/KeyboardSettings/main.cpp:58]: Why unknown if it's in `rc`?

File: Applications/KeyboardSettings/main.cpp [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+        Keyboard::CharacterMap character_map(character_map_file);
+        int ret_val = character_map.set_system_map();
+        if (ret_val != 0) {
+            String message;
+            if (ret_val == -EPERM)
+                message = "Permission denied.";
+            else {
+                message = String::format("Unknown error: %d", ret_val);
+            }
```
Comment (awesomekling) [inline Applications/KeyboardSettings/main.cpp:125]: The error is not unknown, we can get an error string from `strerror(-ret_val)`.

**Pull request merged**